tests/shouldwork/Cores/Xilinx/Floating/TH.hs

```diff
@@ -536,6 +522,30 @@ nanTest =
     , (nan, nan, F.xilinxNaN)
     ]
 
+fromUBasicSamples :: [(Unsigned 32, Float)]
+fromUBasicSamples = delayOutput0 (natToNum @F.FromU32DefDelay) $
+  map (\x -> (x, fromIntegral x))
+    [ 0
+    , 1
+    , maxBound
+    , 0b0010_1010_1010_1010_1010_1010_1010_1010
+    , 0b0101_0101_0101_0101_0101_0101_0101_0101
+    , 0b1010_1010_1010_1010_1010_1010_1010_1010
+    , 0b1101_0101_0101_0101_0101_0101_0101_0101
```

Comment:
The way `fromIntegral` works here is quite opaque to me! 

```
λ:> fromIntegral (0b0000_0000_1111_1111_1111_1111_1111_1111::Unsigned 32) :: Float
1.6777215e7
λ:> fromIntegral (0b0000_0000_1111_1111_1111_1111_1111_1111::Unsigned 32) :: Int
16777215
```

I wonder if there would be a clearer way to handle floats where we can show bits. Something like

```
u32f :: Word32 -> Float
u32f b = unsafePerformIO $ alloca $ \bytes -> poke (castPtr bytes) b *> peek bytes
```

(the function itself is stupid, but it makes clear which bits you are setting in the `Float` and how it all comes together)

Reply:
I think I don't really understand what you mean or what it is that is vague to you, but I'll take a stab at an answer: are you looking for [`decodeFloat`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Prelude.html#v:decodeFloat)? For NaN, there's also [this](https://github.com/clash-lang/clash-compiler/blob/e0266278ae4f8f242559858e19c94d8f8a0e2be6/tests/shouldwork/Cores/Xilinx/Floating.hs#L29) in this very test, and decoding infinity is simply an application of [`isInfinite`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Prelude.html#v:isInfinite). That just leaves the sign bit. Also, of course, you can just `bitCoerce :: Float -> BitVector 32` if all you want is look at the bits.

I don't understand your `u32f`, I'm not familiar with peeking and poking outside of BASIC on the Commodore 64 ;-).

Reply:
Oh I see! Yes, `encodeFloat` or `bitCoerce` is much better than what I had. 

I guess `fromIntegral` works too, if you prefer.

Reply:
I'm not trying to encode a float, I'm trying to give it an `Unsigned 32` and the expected result of `fromU32` applied to that `Unsigned 32`. The Haskell implementation of `fromU32` is exactly `fromIntegral` followed by a dummy pipeline, and I'm trying to verify in this test whether interesting cases behave the same in Vivado. I'm letting Haskell do all the hard work of converting an integral number into a float, since I'm not trying to verify that Haskell does this correctly but rather that it does it in the same way as Vivado.

Are you confusing `fromIntegral :: (Integral a, Num b) => a -> b` with `bitCoerce`? `fromIntegral` just converts a 1 into a 1.0 and a 2 into a 2.0, but if you give it something which cannot be represented in floating point, it will get rounded.

Reply:
>fromIntegral just converts a 1 into a 1.0 and a 2 into a 2.0

Oh right, that makes a lot of sense, sorry for the diversion.